### PR TITLE
Support using tree structures in PEPS optimizations

### PR DESCRIPTION
**Pull request by LinjianMa**

### CODE REVIEW

**Comment by LinjianMa** [comment]
This one contains unmerged changes in #33, we need to finish reviewing that PR first.

File: src/ITensorNetworks/peps.jl [hunk 1]
```diff
@@ -106,17 +106,167 @@ end
...
+Example
+-------
+   |  |   |
+p1-p2 |  p3-p4
+ | |  |   | |
+ | |  |   | |
+s1-s2-s3-s4-s5
+here line_size=5, center_index=3, si represents the list of tensors returned by site_tensors(i),
+projectors are [p1, p2, p3, p4].
+Returns two trees: [[s1, p1], s2, p2] and [[s5, p4], s4, p3]
```
Comment (mtfishman) [inline src/ITensorNetworks/peps.jl:134]: So the trees this returns already encode the optimal contraction sequence?

Also, in general the structure would contain projector tensors on both sides of the `si` tensors, right?

**Comment by mtfishman** [comment]
There's a lot of sophisticated functionality packed in here. I'm having a bit of trouble distinguishing which functions are doing what, could you give a high level summary of the logic for me?

File: src/ITensorNetworks/peps.jl (hunk 1 above)
Comment (LinjianMa) [reply src/ITensorNetworks/peps.jl:134]: yes that's right, in general projectors on both sides will be included.

Yes I already encode the optimal contraction sequence. The reason is that I find for large systems, cotengra still cannot find optimal path for the 1D structure consisting of a line and accompanying projectors. So finally just did it myself.

**Comment by LinjianMa** [comment]
> There's a lot of sophisticated functionality packed in here. I'm having a bit of trouble distinguishing which functions are doing what, could you give a high level summary of the logic for me?

Yes,
Function `tree` finds the optimal contraction path for 1-D components (a line structure with accompanying projectors)

Function `inner_network` with ::type(tree) argument returns a tree structure representing <peps|H|peps> or <peps|peps>

Function `inner_networks` with ::type(tree) argument returns a batch of inner network, <p|H_1|p>, ..., <p|H_n|p>, each one is a tree structure

Function `neighboring_tensors` returns part of tensor_list that are neighboring to subnetwork

I think these are the new components implemented, let me know if something is unclear

**Comment by mtfishman** [comment]
That makes sense, thanks!

For the `tree` function, that determines the optimal contraction path for the 1D network by hand, right? By which I mean, it is hand-coded instead of determining the optimal contraction path automatically with something like cotengra? So with this design, does it make use of contraction path optimization at all?

**Comment by LinjianMa** [comment]
> That makes sense, thanks!
> 
> For the `tree` function, that determines the optimal contraction path for the 1D network by hand, right? By which I mean, it is hand-coded instead of determining the optimal contraction path automatically with something like cotengra? So with this design, does it make use of contraction path optimization at all?

Yes you are right, for now that is purely determined by hand

File: src/ITensorNetworks/peps.jl (hunk 1 above)
Comment (mtfishman) [reply src/ITensorNetworks/peps.jl:134]: Interesting. I find it strange that it even has trouble with that simple network, I think the more brute force optimization code in this module: https://github.com/ITensor/ITensors.jl/tree/main/src/ContractionSequenceOptimization could determine those networks without too much effort.

Comment (mtfishman) [reply src/ITensorNetworks/peps.jl:134]: But if you already hand-coded it, that is even better. We can work on more automation later.

**Comment by mtfishman** [comment]
I don't follow all of the code, but the design seems reasonable. We can merge it and circle back to things like making it more automated  in the future.

**Pull request merged**